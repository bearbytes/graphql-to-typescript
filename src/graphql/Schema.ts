import axios from 'axios'
import {
  getIntrospectionQuery,
  buildClientSchema,
  IntrospectionQuery,
  buildSchema,
} from 'graphql/utilities'
import { GraphQLSchema } from 'graphql'
import { readFileSync, writeFileSync } from 'fs'

export async function downloadSchema(url: string): Promise<GraphQLSchema> {
  const query = getIntrospectionQuery({ descriptions: false })
  const data = { query }
  const response = await axios.post(url, data)
  const introspection = response.data.data as IntrospectionQuery
  const schema = buildClientSchema(introspection, { assumeValid: true })
  return schema
}

export function loadSchema(filePath: string): GraphQLSchema {
  return buildClientSchema(
    JSON.parse(readFileSync(filePath, { encoding: 'utf8' }))
  )
}

export function saveSchema(schema: GraphQLSchema, filePath: string) {
  writeFileSync(filePath, JSON.stringify(schema, null, 2))
}
