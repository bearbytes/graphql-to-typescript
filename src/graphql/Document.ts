import fs from 'fs-extra'
import { parse, DocumentNode } from 'graphql'

export async function loadDocument(filepath: string): Promise<DocumentNode> {
  try {
    const fileContent = await fs.readFile(filepath, { encoding: 'utf8' })
    return parse(fileContent)
  } catch (e) {
    throw { message: `Error parsing document at ${filepath}: ${e.message}` }
  }
}
