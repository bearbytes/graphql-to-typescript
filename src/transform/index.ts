import {
  GraphQLSchema,
  GraphQLNamedType,
  isScalarType,
  GraphQLScalarType,
  isObjectType,
  GraphQLObjectType,
  GraphQLInputObjectType,
  GraphQLEnumType,
  GraphQLInterfaceType,
  GraphQLUnionType,
  isInterfaceType,
  isUnionType,
  isEnumType,
  isInputObjectType,
  GraphQLType,
} from 'graphql'
import { values, compose } from 'rambda'

interface SchemaIR {
  types: NamedTypeIR[]
}
export default function transformSchema(schema: GraphQLSchema): SchemaIR {
  const types = values(schema.getTypeMap())
    .filter((T) => !T.name.startsWith('__'))
    .map(transformNamedType)

  return { types }
}

type NamedTypeIR = TypeIR & {
  typename: string
}
function transformNamedType(T: GraphQLNamedType): NamedTypeIR {
  const typename = T.name
  if (typename.startsWith('__')) return

  const type = transformType(T)
  return { ...type, typename }
}

type TypeIR =
  | ScalarTypeIR
  | ObjectTypeIR
  | InterfaceTypeIR
  | UnionTypeIR
  | EnumTypeIR
  | InputObjectTypeIR
function transformType(T: GraphQLType): TypeIR {
  if (isScalarType(T)) {
    return transformScalarType(T)
  }
  if (isObjectType(T)) {
    return transformObjectType(T)
  }
  if (isInterfaceType(T)) {
    return transformInterfaceType(T)
  }
  if (isUnionType(T)) {
    return transformUnionType(T)
  }
  if (isEnumType(T)) {
    return transformEnumType(T)
  }
  if (isInputObjectType(T)) {
    return transformInputObjectType(T)
  }
}

interface ScalarTypeIR {
  kind: 'scalar'
  scalar: string
}
function transformScalarType(T: GraphQLScalarType): ScalarTypeIR {
  return {
    kind: 'scalar',
    scalar: T.name,
  }
}

interface ObjectTypeIR {
  kind: 'object'
}
function transformObjectType(T: GraphQLObjectType): ObjectTypeIR {
  return {
    kind: 'object',
  }
}

interface InterfaceTypeIR {
  kind: 'interface'
}
function transformInterfaceType(T: GraphQLInterfaceType): InterfaceTypeIR {
  return {
    kind: 'interface',
  }
}

interface UnionTypeIR {
  kind: 'union'
}
function transformUnionType(T: GraphQLUnionType): UnionTypeIR {
  return { kind: 'union' }
}

interface EnumTypeIR {
  kind: 'enum'
}
function transformEnumType(T: GraphQLEnumType): EnumTypeIR {
  return { kind: 'enum' }
}

interface InputObjectTypeIR {
  kind: 'inputObject'
}
function transformInputObjectType(
  T: GraphQLInputObjectType
): InputObjectTypeIR {
  return { kind: 'inputObject' }
}
