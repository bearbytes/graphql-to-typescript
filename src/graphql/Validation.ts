import { validate, GraphQLSchema, DocumentNode } from 'graphql'

export function validateDocument(
  document: DocumentNode,
  schema: GraphQLSchema
) {
  const validationErrors = validate(schema, document)
  for (const e of validationErrors) {
    throw { message: `Error validating document: ${e.message}` }
  }
}
