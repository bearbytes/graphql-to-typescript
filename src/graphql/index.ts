import axios from 'axios'
import {
  getIntrospectionQuery,
  buildClientSchema,
  IntrospectionQuery,
} from 'graphql/utilities'
import { GraphQLSchema } from 'graphql'

export async function parseFromUrl(url: string): Promise<GraphQLSchema> {
  const query = getIntrospectionQuery({ descriptions: false })
  const data = { query }
  const response = await axios.post(url, data)
  const introspection = response.data.data as IntrospectionQuery
  const schema = buildClientSchema(introspection, { assumeValid: true })
  return schema
}
