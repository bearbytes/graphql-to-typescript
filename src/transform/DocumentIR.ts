import { DocumentNode, DefinitionNode } from 'graphql'
import { zipObj } from 'rambda'
import { OperationIR, transformOperation } from './OperationIR'

export interface DocumentIR {
  operations: Record<string, OperationIR>
}

export function transformDocument(T: DocumentNode): DocumentIR {
  const ops = T.definitions.map(transformDefinition)

  const operations = zipObj(ops.map((it) => it.kind), ops)

  return { operations }
}

export function transformDefinition(T: DefinitionNode): OperationIR {
  switch (T.kind) {
    case 'OperationDefinition':
      return transformOperation(T)
    default: {
      console.warn(`Unhandled DefinitionNode.kind '${T.kind}'`)
    }
  }
}
