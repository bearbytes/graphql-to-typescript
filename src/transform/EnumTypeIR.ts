import { GraphQLEnumType, GraphQLEnumValue } from 'graphql'

export interface EnumTypeIR {
  kind: 'enum'
  values: string[]
}

export function transformEnumType(T: GraphQLEnumType): EnumTypeIR {
  return { kind: 'enum', values: T.getValues().map(enumTypeName) }
}

function enumTypeName(T: GraphQLEnumValue) {
  return T.name
}
