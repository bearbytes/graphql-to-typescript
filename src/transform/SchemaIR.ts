import { GraphQLSchema } from 'graphql'
import { values, zipObj, keys } from 'rambda'
import { NamedTypeIR, transformNamedType } from './NamedTypeIR'

export interface SchemaIR {
  types: Record<string, NamedTypeIR>
}

export function transformSchema(schema: GraphQLSchema): SchemaIR {
  const types = zipObj(
    keys(schema.getTypeMap()) as string[],
    values(schema.getTypeMap())
      .filter((T) => !T.name.startsWith('__'))
      .map(transformNamedType)
  )

  return { types }
}
