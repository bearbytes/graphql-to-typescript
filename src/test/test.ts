import { loadSchema } from '../graphql/Schema'
import { transformSchema } from '../transform/SchemaIR'

test('parse schema', async () => {
  const schema = await loadSchema('src/test/schema.gql')
  expect(schema).toMatchSnapshot('Schema')

  const schemaIR = transformSchema(schema)
  expect(schemaIR).toMatchSnapshot('SchemaIR')
})
