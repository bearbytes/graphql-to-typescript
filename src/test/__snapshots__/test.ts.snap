// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse schema: Schema 1`] = `
GraphQLSchema {
  "__allowedLegacyNames": Array [],
  "__validationErrors": undefined,
  "_directives": Array [
    "@skip",
    "@include",
    "@deprecated",
  ],
  "_implementations": Object {
    "UnionType": Array [],
  },
  "_mutationType": "Mutation",
  "_possibleTypeMap": Object {},
  "_queryType": "Query",
  "_subscriptionType": null,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "CustomScalar": "CustomScalar",
    "Float": "Float",
    "ID": "ID",
    "InputType": "InputType",
    "Int": "Int",
    "Mutation": "Mutation",
    "NestedInputType": "NestedInputType",
    "NestedObjectType": "NestedObjectType",
    "ObjectType": "ObjectType",
    "Potato": "Potato",
    "Query": "Query",
    "String": "String",
    "Tomato": "Tomato",
    "UnionType": "UnionType",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": undefined,
  "extensionASTNodes": undefined,
}
`;

exports[`parse schema: SchemaIR 1`] = `
Object {
  "types": Object {
    "Boolean": Object {
      "kind": "namedType",
      "typename": "Boolean",
    },
    "CustomScalar": Object {
      "kind": "namedType",
      "typename": "CustomScalar",
    },
    "Float": Object {
      "kind": "namedType",
      "typename": "Float",
    },
    "ID": Object {
      "kind": "namedType",
      "typename": "ID",
    },
    "InputType": Object {
      "kind": "namedType",
      "typename": "InputType",
    },
    "Int": Object {
      "kind": "namedType",
      "typename": "Int",
    },
    "Mutation": Object {
      "kind": "namedType",
      "typename": "Mutation",
    },
    "NestedInputType": Object {
      "kind": "namedType",
      "typename": "NestedInputType",
    },
    "NestedObjectType": Object {
      "kind": "namedType",
      "typename": "NestedObjectType",
    },
    "ObjectType": Object {
      "kind": "namedType",
      "typename": "ObjectType",
    },
    "Potato": Object {
      "kind": "namedType",
      "typename": "Potato",
    },
    "Query": Object {
      "kind": "namedType",
      "typename": "Query",
    },
    "String": Object {
      "kind": "namedType",
      "typename": "String",
    },
    "Tomato": Object {
      "kind": "namedType",
      "typename": "Tomato",
    },
    "UnionType": Object {
      "kind": "namedType",
      "typename": "UnionType",
    },
    "__Directive": undefined,
    "__DirectiveLocation": undefined,
    "__EnumValue": undefined,
    "__Field": undefined,
    "__InputValue": undefined,
    "__Schema": undefined,
    "__Type": undefined,
    "__TypeKind": undefined,
  },
}
`;
